### ISSUE TITLE
Function.prototype.bind definition interferes with other javascript libraries

### ISSUE THREAD

**Issue by notabeatle** [jagthedrummer/codemirror-ui#8]
The Function.prototype.bind definition in the last few lines of codemirror-ui doesn't play nicely with the Jabber/XMPP library strophe.js (kills connection attempts very early in the process), and, I strongly suspect, plenty of other libraries and codebases, especially those of the non-jQuery variety (all of the dozen or so other libraries I am using are jQuery based, and seem to be unaffected, either by fortune or by some sort of insulating effect created by that library).

My early guess at a fix would be to somehow scope this behavior so it doesn't affect functions declared outside codemirror-ui.  Not knowing _precisely_ how and where the library relies on it makes working around or fixing it difficult.  Perhaps creating a subclass of the Function object, setting its .prototype.bind, then subclassing any functions that require it from that rather than directly from Function would do the trick, though if codemirror-ui relies on this behavior from built-in methods in addition to its own then that may not be enough.

EDIT: Manually adding the code from that bottom method declaration to _every single_ function(){} in the file (changing it to this.bind rather than prototype.bind) and reverting an apparent change near (IIRC) line 283 (by the numbering before I modified the file) seems to have fixed it, though that is a very, _very_ ugly way to do it.

If anyone desperately needs to use this alongside strophe.js or any other code that assumes base classes like Function haven't been modified (tsk tsk!) ask and I can post it, but otherwise it'd be better to wait for it to be fixed by someone who's both better at javascript and more familiar with this code than I am.

The snippet I added to all functions was:

``` javascript
this.bind = function(scope) {
  var _function = this;

  return function() {
    return _function.apply(scope, arguments);
  }
}
```

The bit I commented/uncommented was this:

``` javascript
//event.target.func();
//return false;
this.self[action].call(this);
eval("this."+action)();
```

Commented out lines 1 & 2, uncommented 3 & 4.

**Comment by jagthedrummer** [comment]
Thanks for the report and detailed info.  Any chance you could put together a small example that demonstrates the problem?

**Comment by notabeatle** [comment]
1. Download the main strophe.js package and extract it.  It includes an examples directory; I modified basic.html for this demo, but I'd guess any of them would do.
2. Load basic.html in a browser (it may have to actually be on a remote server due to restrictions on local javascript in some browsers, but I'm not certain of that) and try it out.  If you've got a Jabber server with BOSH enabled, great; if not, just do what I did (this time—when I made the initial report I _did_ have a Jabber server set up and working with HTTP transport) and use a gmail account for the login (including the @gmail.com).  It'll fail, but the important part is that you get several lines of output as a full attempt is made to communicate with the server.
3. Download the latest codemirror and codemirror-ui libs.  Put them somewhere that's accessible to basic.html.
4. Modify the header section of basic.html so it looks something like this:

```
    <script src='https://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js'></script>
    <script src="lib/codemirror.js"></script>
    <script src="js/codemirror-ui.js"></script>   
    <script src='../strophe.js'></script>
    <script src='basic.js'></script>
```
1. Load the page and try it again.  This time you'll get just one line of output, plus a bunch of type errors in your JS console if you have it open.
2. Comment out the line that includes codemirror-ui, but leave codemirror itself in.  Boom, strophe.js works again.

To be fair, I think Strophe might be modifying Function.prototype as well&mdash;however, IMO _neither_ should to avoid exactly this sort of problem.

This time I tested in Chrome, but, as noted in my issue report above, as of Firefox 5 that browser displayed the same behavior and I doubt that's changed.  My guess would be that this issue is browser-independent.

**Comment by jagthedrummer** [comment]
Awesome, thanks for the detailed reproduction steps.  I'll check this out and figure out how to isolate CMUI from the other libraries.

Thanks,
Jeremy

**Comment by jagthedrummer** [comment]
OK, so I figured out that it was a collision between the 'bind' method in CMUI and a different 'bind' method in strophe.  I just changed the name of my bind function to 'cmuiBind' and everything started working.  I figure that's enough of a 'namespace' to keep that sort of direct collision from being a problem with other libs.  I probably should figure out a better long term fix, just so I'm not polluting the main Function prototype.  Thanks for pointing this out.  I just released version 0.0.19 that contains this fix.  (Note that there's a new version of CodeMirror included now, and the include block has changed slightly.)

Thanks,
Jeremy
